### PR TITLE
zpotrf_dtd fixed for the GPU, updated kernels to cublas_v2, although results still suspicious

### PR DESCRIPTION
**Pull request by BrieucNicolas**
testing_zpotrf_dtd now works with GPU. However, the factorisation and results produced are suspicious

### CODE REVIEW

**Comment by BrieucNicolas** [comment]
This is the continuation of PR #55

File: src/include/dplasma/constants.h [hunk 1]
```diff
@@ -209,4 +209,85 @@ enum dplasma_matrix_type_e {
 extern char *dplasma_lapack_const_strings[];
 #define dplasma_lapack_const(plasma_const) (dplasma_lapack_const_strings[plasma_const][0])
 
+#if defined(DPLASMA_HAVE_CUDA)
+#include <cublas.h>
```
[the remark this answers is not in the document]
Comment (BrieucNicolas) [reply src/include/dplasma/constants.h:213]: Yes, it depends on the order of inclusion of the headers. In order not to break the rest of dplasma, it stays cublas.h here but cublas_v2 is included elsewhere and this should compile ?

File: src/include/dplasma/constants.h [hunk 2]
```diff
@@ -209,4 +209,85 @@ enum dplasma_matrix_type_e {
...
+dplasma_cublas_side(dplasma_enum_t* side)
+{
+    assert( (*side == dplasmaRight) || (*side == dplasmaLeft) );
+    switch(*side){
+        case dplasmaRight:
+            *side = CUBLAS_SIDE_RIGHT;
+            break;
+        case dplasmaLeft:
+            *side = CUBLAS_SIDE_LEFT;
+            break;
```
[the remark this answers is not in the document]
Comment (BrieucNicolas) [reply src/include/dplasma/constants.h:225]: Should i make them all macros like so ?

`#define dplasma_cublas_diag(diag)
    assert( (*diag == dplasmaNonUnit) || (*diag == dplasmaUnit) );
    *diag == dplasmaNonUnit ? *diag = CUBLAS_DIAG_NON_UNIT : *diag = CUBLAS_DIAG_UNIT;
`

It's more consistent with `dplasma_lapack_const` too

File: src/include/dplasma/constants.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (QingleiCao) [reply src/include/dplasma/constants.h:213]: I see. We had issues using CUDA12: [#92](https://github.com/ICLDisco/dplasma/issues/92). We can open another PR for that to upgrade from `cublas` to `cublas_v2`.

File: tests/testing_zpotrf_dtd.c [hunk 3]
```diff
@@ -105,98 +115,270 @@ parsec_core_gemm(parsec_execution_stream_t *es, parsec_task_t *this_task)
...
+        cublas_uplo = CUBLAS_FILL_MODE_UPPER;
+
+    status = cusolverDnCreate(&cusolverDnHandle);
+    assert(CUSOLVER_STATUS_SUCCESS == status);
+    (void)status;
+
+    status = cusolverDnDpotrf_bufferSize(cusolverDnHandle, cublas_uplo, nb, NULL, mb, &workspace_size);
+    assert(CUSOLVER_STATUS_SUCCESS == status);
+
+    cusolverDnDestroy(cusolverDnHandle);
```
[the remark this answers is not in the document]
Comment (bosilca) [reply tests/testing_zpotrf_dtd.c]: Technically each stream needs a handle, but the handle creation is a function in userland. Unfortunately, we do not expose the streams to the users so these need to be done as needed via a function exported by the user code.

File: tests/common.c [hunk 4]
```diff
@@ -716,9 +721,7 @@ parsec_context_t* setup_parsec(int argc, char **argv, int *iparam)
         }
     }
     if( nbgpu > 0 ) {
-        cublasStatus_t status = cublasInit();
-        assert(CUBLAS_STATUS_SUCCESS == status);
-        parsec_info_register(&parsec_per_stream_infos, "DPLASMA::CUDA::HANDLES",
+        CuHI = parsec_info_register(&parsec_per_stream_infos, "DPLASMA::CUDA::HANDLES",
                              destroy_cuda_handles, NULL,
```
Comment (bosilca) [inline tests/common.c:725]: why one of these function is name-spaced with dplasma while the other one is not ? Why is the info called `parsec_per_stream_infos` when it clearly belongs to dplasma. I know there issues are not related to this PR, but I just noticed them while reading through the code.

File: tests/common_dtd_z.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+int parsec_core_ztrsm_cuda( parsec_device_gpu_module_t* gpu_device,
+                            parsec_gpu_task_t*          gpu_task,
+                            parsec_gpu_exec_stream_t*   gpu_stream);
+int parsec_core_zgemm_cuda( parsec_device_gpu_module_t* gpu_device,
+                            parsec_gpu_task_t*          gpu_task,
+                            parsec_gpu_exec_stream_t*   gpu_stream);
+
+#endif /* DPLASMA_HAVE_CUDA */
+
+#define PASTE_CODE_CREATE_ZPOTRF_TASKCLASS(TASKPOOL, TASKCLASS)                                                        \
```
Comment (bosilca) [inline tests/common_dtd_z.h:62]: I don't think these are doing any good, they make the code look simpler but less understandable. Readers need to see what they are doing, not have obscure macro sprinkled across the code, especially if there are very few uses.

File: tools/PrecisionGenerator/subs.py [hunk 6]
```diff
@@ -297,6 +297,7 @@
     ('twoDBC_s',    'twoDBC_d',    'twoDBC_c',    'twoDBC_z'   ),
     ('workspace_s', 'workspace_d', 'workspace_c', 'workspace_z'),
     ('dplasmacore_s',   'dplasmacore_d',   'dplasmacore_c',   'dplasmacore_z'  ),
+    ('dtd_s',       'dtd_d',       'dtd_c',       'dtd_z'      ),
```
Comment (bosilca) [inline tools/PrecisionGenerator/subs.py:300]: Are we sure this is necessary ?

File: src/dplasmaaux.h [hunk 7]
```diff
@@ -110,14 +110,35 @@ extern void *dplasma_pcomm;
...
+    do {                                                                                  \
+        cusolverStatus_t __cusolver_status = (cusolverStatus_t) (STATUS);                 \
+        if( CUSOLVER_STATUS_SUCCESS != __cusolver_status ) {                              \
+            parsec_warning( "%s:%d %s%s", __FILE__, __LINE__,                             \
+                            (STR), dplasma_cusolver_error_to_string(__cusolver_status) ); \
+            CODE;                                                                         \
+        }                                                                                 \
+    } while(0)
+
 #endif
```
[the remark this answers is not in the document]
Comment (BrieucNicolas) [reply src/dplasmaaux.h:142]: I don't know about all the types but `dplasma_potrf_workspace_t`is used by zpotrf_wrapper, and my test when creating, destroying and acessing the workspace.

**Review by bosilca — APPROVED** [review]

**Review by bosilca — APPROVED** [review]

**Review by bosilca — APPROVED** [review]

**Pull request merged**